### PR TITLE
Feature/validate instance

### PR DESCRIPTION
**Pull request by alexmilde**
Add option to validate an instance.
Usefull if validation shall be done before submitting the whole form.

### CODE REVIEW

**Comment by alexmilde** [comment]
hi patrick,
thanks for your amazing lib. 

i was missing a specific feature so i added it: 
the option to validate an instance via public method.

as i'm not an open source contribution pro: don't hesitate to comment on my proposed change.

cheers and thanks for your work.

**Comment by patrickkunka** [comment]
Thanks for this @alexmilde! Appreciate the contribution. I will review properly in the next few days and hopefully get it merged, but at a glance it looks good.

**Comment by alexmilde** [comment]
hey patrick,
did you have the time to look into it?
happy new year btw!

File: README.md [hunk 1]
```diff
@@ -716,6 +716,19 @@ const edd = easydropdown(selectElement);
...
+
+`.validate()`
+
+Validates the instance and returns a boolean whether validation succeeded. Helpful if you need to validate without hitting the form submit button. 
+
+```js
+
+const edd = easydropdown(selectElement);
+
+edd.validate();
```
Comment (patrickkunka) [inline README.md]: Please extend the example to demonstrate how we can use the returned boolean, e.g:

```js
const isValid = edd.validate();

console.log(isValid); // false
```

File: demos/15-callbacks.html [hunk 2]
```diff
@@ -19,8 +19,8 @@ <h2 class="sub-heading">EasyDropDown Demo</h2>
             <h1 class="heading">15. Callbacks</h1>
 
             <div class="demo-container">
-                <select id="test">
-                    <option value="" data-placeholder>Select</option>
+                <select id="test" required>
```
Comment (patrickkunka) [inline demos/15-callbacks.html]: Can you revert the changes to this demo file? I'm guessing you didn't intend to commit them?

File: src/Easydropdown/Easydropdown.test.ts [hunk 3]
```diff
@@ -152,4 +152,30 @@ describe('Easydropdown', () => {
...
         });
     });
+
+    describe('.validate()', () => { 
+
+        it('validates a required select', () => {
+            const select = createSelect('required');
+            const edd = new Easydropdown(select, {});
+
+            // @ts-ignore
```
Comment (patrickkunka) [inline src/Easydropdown/Easydropdown.test.ts]: don't think you need these @ts-ignore - in the other tests, they were needed so we can test private methods, but validate is part of the public API so should be fine.

File: src/Easydropdown/Easydropdown.test.ts [hunk 4]
```diff
@@ -6,11 +6,11 @@ import Renderer from '../Renderer/Renderer';
 
 import Easydropdown  from './Easydropdown';
 
-const createSelect = () => {
+const createSelect = (selectAttributes = '') => {
```
Comment (patrickkunka) [inline src/Easydropdown/Easydropdown.test.ts]: let's make this an array, it will read better:

```ts
const createSelect = (selectAttributes: string[] = []) => {
   ...
   <select ${selectAttributes.join(' ')}>
   ...
}
```

File: src/Easydropdown/Easydropdown.test.ts [hunk 5]
```diff
@@ -152,4 +152,30 @@ describe('Easydropdown', () => {
...
+        it('validates a non required select', () => {
+            const select = createSelect();
+            const edd = new Easydropdown(select, {});
+
+            // @ts-ignore
+            assert.equal(edd.validate(), true);
+            edd.actions.selectOption(1);
+            // @ts-ignore
+            assert.equal(edd.validate(), true);    
+        });
```
Comment (patrickkunka) [inline src/Easydropdown/Easydropdown.test.ts]: My linter is saying you have some trailing whitespace around these blocks - are you able remove it? (vscode has an option to auto remove on save if you're using vscode)

File: src/Easydropdown/Easydropdown.test.ts [hunk 6]
```diff
@@ -152,4 +152,30 @@ describe('Easydropdown', () => {
             assert.equal(edd.dom.item.length, 7);
         });
     });
+
+    describe('.validate()', () => {
```
Comment (patrickkunka) [inline src/Easydropdown/Easydropdown.test.ts]: Please create two additional test cases for a required and non-required select with a "placeholder" option. See my comments further down for an explanation.

File: src/Easydropdown/Easydropdown.test.ts [hunk 7]
```diff
@@ -152,4 +152,30 @@ describe('Easydropdown', () => {
...
     });
+
+    describe('.validate()', () => { 
+
+        it('validates a required select', () => {
+            const select = createSelect('required');
+            const edd = new Easydropdown(select, {});
+
+            // @ts-ignore
+            assert.equal(edd.validate(), false);
```
Comment (patrickkunka) [inline src/Easydropdown/Easydropdown.test.ts:169]: This test is currently failing - see my comment in `Easydropdown.ts` as to why.

File: src/Easydropdown/Easydropdown.ts [hunk 8]
```diff
@@ -99,6 +99,16 @@ class Easydropdown {
         Renderer.queryDomRefs(this.dom, ['group', 'option', 'item']);
     }
 
+    public validate(): boolean {
+        if (this.state.selectedIndex !== 0 || this.state.isRequired !== true) {
```
Comment (patrickkunka) [inline src/Easydropdown/Easydropdown.ts]: This logic will only work if the select has a "placeholder" option, i.e. a first option with an empty value attribute, specifically to be used as an initial placeholder/label.

If there is no placeholder, then the selectedIndex will be `0` by default, but the dropdown will still be valid, because the first option will have a value.

There is a `hasValue` getter in the state that is nicely agnostic to whether a placeholder exists, and simply checks if the value is empty string, so I would recommend the following instead:

```ts
public validate(): boolean {
    if (!this.state.isRequired || this.state.hasValue) {
        return true;
    }

    ...
}
```

Note that I've moved the `isRequired` lookup first as I feel it makes sense to avoid the more expensive getter computation if no validation is required.

File: src/State/resolveActions.ts [hunk 9]
```diff
@@ -88,7 +88,7 @@ const resolveActions = (state: State): IActions => ({
 
         state.selectedIndex = index;
 
-        if (state.isInvalid) {
+        if (state.isInvalid && state.selectedIndex !== 0) {
```
Comment (patrickkunka) [inline src/State/resolveActions.ts]: as above, please change to:

```ts
if (state.isInvalid && state.hasValue) {
    this.validate();
}
```

**Review by patrickkunka — CHANGES_REQUESTED** [review]
Hi @alexmilde - happy new year to you too and thanks for your effort on this. I've just had a chance to review and test. Your approach is sound and makes sense but there is one issue that you might have overlooked, please see my comments. Just a few simple changes required.

File: demos/15-callbacks.html (hunk 2 above)
Comment (alexmilde) [reply demos/15-callbacks.html]: Actually i did this on purpose to demonstrate validation on close callback. (line 56) 
shall i revert the whole file? right now i readded the data-placeholder attribute but kept my required

File: src/Easydropdown/Easydropdown.test.ts (hunk 5 above)
Comment (alexmilde) [reply src/Easydropdown/Easydropdown.test.ts]: enabled it but cant see any differnce. would be nice if you could recheck.

**Comment by alexmilde** [comment]
Thanks for your feedback. I was unshure whether you want me to resolve the discussion or you want to. So i left them open.

File: demos/15-callbacks.html (hunk 2 above)
Comment (patrickkunka) [reply demos/15-callbacks.html]: Ah i see - i was confused because the callbacks demo doesn't really relate to validation, so wondered if you'd just added there for your own testing. I think it will be better if we create a new demo for programmatic validation - but I'm happy to do that :) So let's revert this.

File: src/Easydropdown/Easydropdown.test.ts (hunk 5 above)
Comment (patrickkunka) [reply src/Easydropdown/Easydropdown.test.ts]: Yeah all looks good, thanks!

**Comment by patrickkunka** [comment]
Thanks for all the work Alex. Just one thing remaining then I'm happy to merge.

**Comment by patrickkunka** [comment]
FYI - i've created a `v4-develop` branch and re-based the PR to merge against that. I will prepare a release from there once the PR is merged.

File: demos/15-callbacks.html (hunk 2 above)
Comment (alexmilde) [reply demos/15-callbacks.html]: sounds good. thanks.

**Review by patrickkunka — APPROVED** [review]
🎉

**Comment by patrickkunka** [comment]
Hi @alexmilde - FYI - this has just been released in https://github.com/patrickkunka/easydropdown/releases/tag/v4.2.0 thanks for your contribution :)
